relay-sampling/src/lib.rs

```diff
@@ -427,57 +399,95 @@ impl SamplingRule {
+                decayed_sample_rate,
+            } => {
+                if let TimeRange {
+                    start: Some(start),
+                    end: Some(end),
+                } = self.time_range
+                {
+                    if self.sample_rate > decayed_sample_rate && start < end && now >= start {
+                        return Some(MatchingRule {
+                            id: self.id,
+                            evaluator: SampleRateEvaluator::Linear {
+                                start,
+                                end,
+                                sample_rate: self.sample_rate,
+                                decayed_sample_rate,
+                            },
+                        });
+                    }
+                }
+            }
+            DecayingFunction::Constant => {
+                if self.time_range.contains(now) {
+                    return Some(MatchingRule {
+                        id: self.id,
+                        evaluator: SampleRateEvaluator::Constant {
+                            sample_rate: self.sample_rate,
+                        },
+                    });
+                }
+            }
         }
-    }
 
-    /// Validates and gets the decaying function params.
-    fn get_decaying_function_params(&self, now: DateTime<Utc>) -> Option<DecayingFunctionParams> {
-        if let TimeRange {
-            start: Some(start),
-            end: Some(end),
-        } = self.time_range
-        {
-            if !self.validate_function_specific_params() || end <= start || now < start {
-                return None;
-            }
+        None
+    }
+}
 
-            let now_timestamp = now.timestamp() as f64;
-            let start_timestamp = start.timestamp() as f64;
-            let end_timestamp = end.timestamp() as f64;
-            let progress_ratio = ((now_timestamp - start_timestamp)
-                / (end_timestamp - start_timestamp))
-                .clamp(0.0, 1.0);
+/// A sampling rule that has been successfully matched and that contains all the required data
+/// to return the sample rate.
+#[derive(Debug, Clone, Copy)]
+pub struct MatchingRule {
+    pub id: RuleId,
+    evaluator: SampleRateEvaluator,
+}
 
-            Some(DecayingFunctionParams {
-                sample_rate: self.sample_rate,
-                progress_ratio,
-            })
-        } else {
-            None
-        }
+impl MatchingRule {
+    /// Gets the sample rate for the specific rule.
+    pub fn get_sample_rate(&self, now: DateTime<Utc>) -> f64 {
+        self.evaluator.evaluate(now)
     }
+}
 
-    /// Validates the external params specifically to each decaying function.
-    fn validate_function_specific_params(&self) -> bool {
-        match self.decaying_fn {
-            DecayingFunction::Linear {
+/// A struct representing the evaluation context of a sample rate.
+#[derive(Debug, Clone, Copy)]
+enum SampleRateEvaluator {
```

Comment:
We usually declare types before using them, so In this case, declare `SampleRateEvaluator`, then `MatchingRule`, then `SamplingRule`. See [internal docs](https://www.notion.so/sentry/HOWTO-Code-Rust-at-Sentry-7b35f165b10b4492bb95ebe1471a9ada#2695321713704369bccc5fa83ebedfa8).

Reply:
Thanks!